homeassistant/components/smappee/config_flow.py

```diff
@@ -21,3 +25,122 @@ class SmappeeFlowHandler(
+            # We currently only support Energy and Solar models (legacy)
+            return self.async_abort(reason="invalid_mdns")
+
+        serial_number = (
+            discovery_info[CONF_HOSTNAME].replace(".local.", "").replace("Smappee", "")
+        )
+
+        # Check if already configured
+        await self.async_set_unique_id(serial_number)
+        self._abort_if_unique_id_configured()
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update(
+            {CONF_IP_ADDRESS: discovery_info["host"], CONF_SERIALNUMBER: serial_number}
+        )
+
+        return await self.async_step_zeroconf_confirm()
+
+    async def async_step_zeroconf_confirm(self, user_input=None):
+        """Confirm zeroconf flow."""
+        errors = {}
+
+        if user_input is None:
+            # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+            serialnumber = self.context.get(CONF_SERIALNUMBER)
+            return self.async_show_form(
+                step_id="zeroconf_confirm",
+                description_placeholders={"serialnumber": serialnumber},
+                errors=errors,
+            )
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        ip_address = self.context.get(CONF_IP_ADDRESS)
+        serial_number = self.context.get(CONF_SERIALNUMBER)
+
+        # Attempt to make a connection to the local device
+        smappee_api = api.api.SmappeeLocalApi(ip=ip_address)
+        logon = await self.hass.async_add_executor_job(smappee_api.logon)
+        if logon is None:
+            return self.async_abort(reason="connection_error")
+
+        return self.async_create_entry(
+            title=f"Smappee{serial_number}",
+            data={CONF_IP_ADDRESS: ip_address, CONF_SERIALNUMBER: serial_number},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        return await self.async_step_environment()
+
+    async def async_step_environment(self, user_input=None):
+        """Decide environment, cloud or local."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="environment",
+                data_schema=vol.Schema(
+                    {
+                        vol.Required("environment", default=ENV_CLOUD): vol.In(
+                            [ENV_CLOUD, ENV_LOCAL]
+                        )
+                    }
+                ),
+                errors={},
+            )
+
+        # Environment chosen, request additional host information for LOCAL or OAuth2 flow for CLOUD
+        # Ask for host detail
+        if user_input["environment"] == ENV_LOCAL:
+            return await self.async_step_local()
+
+        # Use configuration.yaml CLOUD setup if not setup already
+        for entry in self._async_current_entries():
+            if entry.unique_id.startswith("SmappeeCloud"):
```

Comment:
We may need to guard for `entry.unique_id` being `None` too. Not sure if it could be possible due to race between setting up the existing entries and a new flow.